File: packages/sdk/src/migrations/migrateSnapshot.ts
import { Snapshot } from '@towns-protocol/proto'
import { snapshotMigration0000 } from './snapshotMigration0000'
import { snapshotMigration0001 } from './snapshotMigration0001'
import { snapshotMigration0002 } from './snapshotMigration0002'
import { snapshotMigration0003 } from './snapshotMigration0003'
import { snapshotMigration0004 } from './snapshotMigration0004'
import { snapshotMigration0005 } from './snapshotMigration0005'
import { snapshotMigration0006 } from './snapshotMigration0006'

const SNAPSHOT_MIGRATIONS = [
    snapshotMigration0000,
    snapshotMigration0001,
    snapshotMigration0002,
    snapshotMigration0003,
    snapshotMigration0004,
    snapshotMigration0005,
    snapshotMigration0006,
]

export function migrateSnapshot(snapshot: Snapshot): Snapshot {
    const currentVersion = SNAPSHOT_MIGRATIONS.length
    if (snapshot.snapshotVersion >= currentVersion) {
        return snapshot
    }
    let result = snapshot
    for (let i: number = snapshot.snapshotVersion; i < currentVersion; i++) {
        result = SNAPSHOT_MIGRATIONS[i](result)
    }
    result.snapshotVersion = currentVersion
    return result
}

File: core/node/events/migrations/migrate_snapshot.go
package migrations

import (
	. "github.com/towns-protocol/towns/core/node/protocol"
)

type migrationFunc func(*Snapshot)

// should be kept in sync with packages/sdk/src/migrations/migrate_snapshot.ts
var MIGRATIONS = []migrationFunc{
	snapshot_migration_0000,
	snapshot_migration_0001,
	snapshot_migration_0002,
	snapshot_migration_0003,
	snapshot_migration_0004,
	snapshot_migration_0005,
	snapshot_migration_0006,
}

func CurrentSnapshotVersion() int32 {
	return int32(len(MIGRATIONS))
}

func MigrateSnapshot(iSnapshot *Snapshot) {
	currentVersion := CurrentSnapshotVersion()
	if iSnapshot.SnapshotVersion >= currentVersion {
		return
	}
	for i := iSnapshot.SnapshotVersion; i < currentVersion; i++ {
		MIGRATIONS[i](iSnapshot)
	}
	iSnapshot.SnapshotVersion = currentVersion
}
